### PR TITLE
[ci]: Fix: Make publishing workflows sequential, and complements to #2427

### PR DESCRIPTION
**Pull request by s8sato**
### Description of the Change

1. Make publishing workflows sequential:
    Whenever `iroha2-base:<tag>` updates, `iroha2:<tag>` also updates on the base image. If `Dockerfile.base` has no change, only `iroha2:<tag>` updates
2. Exclude other `.yml` files than workflows from the PR workflow triggers
3. Refactor workflows, etc

### Issue

- None

### Benefits

1. `iroha2-base:<tag>` never diverges from the history of `iroha2:<tag>`
2. Changes to docker-compose files never trigger PR workflows

### Possible Drawbacks

- None

### CODE REVIEW

File: .github/workflows/iroha2-release-pr.yml [hunk 1]
```diff
@@ -65,6 +65,7 @@ jobs:
         with:
           context: .
           push: false
+          load: true
           tags: hyperledger/iroha2:dev
```
Comment (s8sato) [inline .github/workflows/iroha2-release-pr.yml:65]: Making the locally built `hyperledger/iroha2:dev` image available for subsequent steps

File: .github/workflows/iroha2-dev-pr.yml [hunk 2]
```diff
@@ -18,10 +18,9 @@ concurrency:
 
 env:
   CARGO_TERM_COLOR: always
-  CARGO_INCREMENTAL: 0
```
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:21]: `CARGO_INCREMENTAL=0` is implied by [`Swatinem/rust-cache`](https://github.com/Swatinem/rust-cache)

File: .github/workflows/iroha2-dev-nightly.yml [hunk 3]
```diff
@@ -7,8 +7,8 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - name: Login to DockerHub
-        uses: docker/login-action@v2
+      - uses: docker/setup-buildx-action@v2
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev-nightly.yml:10]: This is a bug. We **should have** buildx set up inside the container.

File: .github/workflows/iroha2-dev-pr-static.yml [hunk 4]
```diff
@@ -17,25 +17,28 @@ concurrency:
...
         if: always()
-      - name: no_std compatibility check
-        run: cargo nono check --package iroha_data_model --no-default-features
+        run: cargo +nightly-2022-08-15 doc --no-deps --quiet --no-default-features
+      - name: Documentation with all features enabled
         if: always()
-      - name: Documentation check
-        run: if cargo doc -q 2>&1 | grep '^'; then exit 1; fi
+        run: cargo +nightly-2022-08-15 doc --no-deps --quiet --all-features
+      - name: no_std compatibility
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev-pr-static.yml:42]: These shouldn't be necessary. We are using conditional compilation where we should instead be using LTO.

File: .github/workflows/iroha2-dev-pr-title.yml [hunk 5]
```diff
@@ -18,7 +18,7 @@ jobs:
           regex: '^\[feature\] #\d+(, #\d+)*: .+$'
       # - name: Add feature label
       #   uses: actions-ecosystem/action-add-labels@v1
-      #   if: ${{ steps.feature-match.outputs.match != '' }}
+      #   if: steps.feature-match.outputs.match != ''
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev-pr-title.yml:21]: I don't see why this change was necessary, but OK.

File: .github/workflows/iroha2-dev-pr-unstable.yml [hunk 6]
```diff
@@ -1,4 +1,4 @@
-name: I2::Dev::Tests (Unstable)
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev-pr-unstable.yml:1]: Actually, let's get rid of this workflow, and move it as a separate `job` inside `I2::Dev::Tests`.

Comment (appetrosyan) [reply .github/workflows/iroha2-dev-pr-unstable.yml:1]: The original reason we had this as a separate workflow was because we couldn't restart the failed jobs, just failed workflows. Now that we can re-run failed jobs, we should merge it back.

File: .github/workflows/iroha2-dev-pr.yml [hunk 7]
```diff
@@ -18,10 +18,9 @@ concurrency:
...
 
 jobs:
-  check:
+  checks:
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev-pr.yml:23]:
```suggestion
  check:
```

File: .github/workflows/iroha2-dev-update_base_coverage.yml [hunk 8]
```diff
@@ -1,4 +1,4 @@
-name: I2::Dev::Update Base Coverage
+name: I2::Dev::UpdateBaseCoverage
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev-update_base_coverage.yml:1]:
```suggestion
name: I2::Dev::BaseCoverage
```

File: .github/workflows/iroha2-dev.yml [hunk 9]
```diff
@@ -91,6 +95,8 @@ jobs:
           path: target/schema
 
   telemetry:
+    # FIXME #2646
```
Comment (appetrosyan) [inline .github/workflows/iroha2-dev.yml:98]: Noted

Comment (appetrosyan) [reply .github/workflows/iroha2-dev.yml:98]: #2646

File: .github/workflows/iroha2-release.yml [hunk 10]
```diff
@@ -1,40 +1,52 @@
...
+    steps:
+      - id: tag
+        if: github.ref == 'refs/heads/iroha2-${{ matrix.tag }}'
+        run: echo "::set-output name=tag::${{ matrix.tag }}"
+
+  dockerhub:
+    needs: tag
+    runs-on: ubuntu-latest
+    container:
+      image: 7272721/i2-ci:nightly
```
Comment (appetrosyan) [inline .github/workflows/iroha2-release.yml:29]: We're getting a little overreliant on `matrix`, which contradicts both the simplicity and the performance gains from using a custom image.

File: telemetry/src/ws.rs [hunk 11]
```diff
@@ -2,7 +2,7 @@
 #![allow(clippy::std_instead_of_core, clippy::std_instead_of_alloc)]
 use std::time::Duration;
 
-use chrono::Utc;
+use chrono::Local;
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply telemetry/src/ws.rs:5]: Good!

**Review by appetrosyan — COMMENTED** [review]
Overall LGTM. We should have a brief discussion and then merge.

File: .github/workflows/iroha2-dev-pr-static.yml (hunk 4 above)
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr-static.yml:42]: Are you suggesting to remove `Documentation with all features enabled` and `no_std compatibility` checks?

File: .github/workflows/iroha2-dev-pr-title.yml (hunk 5 above)
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr-title.yml:21]: I just intended to unify to either `if: expression` or `if: ${{ expression }}`

File: .github/workflows/iroha2-release.yml (hunk 10 above)
Comment (s8sato) [reply .github/workflows/iroha2-release.yml:29]: I agree. It would run as many jobs as the number of combinations in the matrix.
Previously I assumed that when used in `<job_id>.if` it could filter out unrelated jobs, but it turned out the matrix values were not evaluated at that point.
I'll take another way

File: .github/workflows/iroha2-dev-nightly.yml (hunk 3 above)
Comment (s8sato) [reply .github/workflows/iroha2-dev-nightly.yml:10]: I added `docker/setup-buildx-action` not because I saw errors but just because it seemed to be the best practice. I'll just remove it then

File: .github/workflows/iroha2-dev-pr-static.yml (hunk 4 above)
Comment (appetrosyan) [reply .github/workflows/iroha2-dev-pr-static.yml:42]: Just the `--all-features` documentation.

File: .github/workflows/iroha2-dev-pr-title.yml (hunk 5 above)
Comment (appetrosyan) [reply .github/workflows/iroha2-dev-pr-title.yml:21]: Got it. Makes sense.

File: .github/workflows/iroha2-dev-pr-static.yml (hunk 4 above)
Comment (appetrosyan) [reply .github/workflows/iroha2-dev-pr-static.yml:42]: Can be done in a separate PR.

**Review by appetrosyan — DISMISSED** [review]

File: .github/workflows/iroha2-dev-pr-static.yml (hunk 4 above)
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr-static.yml:42]: #2663

Comment (s8sato) [reply .github/workflows/iroha2-dev-pr-static.yml]: Done in this PR

**Review by pesterev — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

File: .github/workflows/iroha2-dev.yml [hunk 12]
```diff
@@ -1,7 +1,9 @@
 name: I2::Dev::Publish
 
 on:
-  push:
+  workflow_run:
+    workflows: [I2::Base::Publish]
+    types: [completed]
     branches: [iroha2-dev]
```
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:7]: This is mis-implementation. The workflow was not triggered after this PR merge

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml:7]: Problems:

- [ ] `workflow_run` event lacks `paths` filters
- [ ] Not sure if `completed` condition reacts to the preceding workflow skipped
- [ ] Use of [`workflow_run`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run) needs to be committed to the default branch
	image

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml:7]: Solution:
- Publish the base image not as a workflow but simply as an additional step to each workflow
